### PR TITLE
Add Android-Universal-Image-Loader Extensions

### PR DESCRIPTION
**Pull request by skyfishjy**
As many projects using  [Android-Universal-Image-Loader](https://github.com/nostra13/Android-Universal-Image-Loader). So I write this extension.

### CODE REVIEW

File: scissors-sample/build.gradle [hunk 1]
```diff
@@ -30,7 +30,9 @@ dependencies {
     compile 'com.android.support:design:' + rootProject.ext.supportVersion
     compile 'com.jakewharton:butterknife:7.0.1'
     compile 'com.squareup.leakcanary:leakcanary-android:1.3.1'
-    compile 'com.squareup.picasso:picasso:2.5.2'
+    compile 'com.nostra13.universalimageloader:universal-image-loader:1.9.4'
+    // Or Picasso
+    // compile 'com.squareup.picasso:picasso:2.5.2'
```
Comment (eveliotc) [inline scissors-sample/build.gradle]: Please leave picasso as default for demo

File: scissors-sample/src/main/java/com/lyft/android/scissorssample/App.java [hunk 2]
```diff
@@ -18,18 +18,21 @@
...
 
+    private RefWatcher refWatcher;
+
     public static RefWatcher getRefWatcher(Context context) {
         App application = (App) context.getApplicationContext();
         return application.refWatcher;
     }
 
-    private RefWatcher refWatcher;
-
```
Comment (eveliotc) [inline scissors-sample/src/main/java/com/lyft/android/scissorssample/App.java:32]: Revert this please

File: scissors-sample/src/main/java/com/lyft/android/scissorssample/App.java [hunk 3]
```diff
@@ -47,5 +50,10 @@ public void onCreate() {
                 .penaltyLog()
                 .penaltyDeath()
                 .build());
+
+        // If using Android-Universal-Image-Loader
+        // Create global configuration and initialize ImageLoader with default config
+        ImageLoaderConfiguration config = new ImageLoaderConfiguration.Builder(this).build();
+        ImageLoader.getInstance().init(config);
```
Comment (eveliotc) [inline scissors-sample/src/main/java/com/lyft/android/scissorssample/App.java]: As picasso is default please comment this out mentioned to enable it for Android-Universal-Image-Loader

File: scissors-sample/src/main/java/com/lyft/android/scissorssample/CropResultActivity.java [hunk 4]
```diff
@@ -54,10 +70,4 @@ protected void onCreate(Bundle savedInstanceState) {
         //        .skipMemoryCache(true)
         //        .into(resultView);
     }
-
-    static void startUsing(File croppedPath, Activity activity) {
-        Intent intent = new Intent(activity, CropResultActivity.class);
-        intent.putExtra(EXTRA_FILE_PATH, croppedPath.getPath());
-        activity.startActivity(intent);
-    }
```
Comment (eveliotc) [inline scissors-sample/src/main/java/com/lyft/android/scissorssample/CropResultActivity.java:62]: Revert this

File: scissors-sample/src/main/java/com/lyft/android/scissorssample/CropResultActivity.java [hunk 5]
```diff
@@ -42,10 +51,17 @@ protected void onCreate(Bundle savedInstanceState) {
...
 
-        Picasso.with(this)
-                .load(imageFile)
-                .memoryPolicy(MemoryPolicy.NO_CACHE, MemoryPolicy.NO_STORE)
-                .into(resultView);
+        DisplayImageOptions options = new DisplayImageOptions.Builder()
+                .cacheInMemory(false)
+                .cacheOnDisk(false)
+                .build();
+        ImageLoader.getInstance().displayImage("file://" + filePath, resultView, options);
```
Comment (eveliotc) [inline scissors-sample/src/main/java/com/lyft/android/scissorssample/CropResultActivity.java]: As mentioned before comment it out leaving picasso only e.g.

`// Or Android-Universal-Image-Loader`

File: scissors-sample/src/main/java/com/lyft/android/scissorssample/MainActivity.java [hunk 6]
```diff
@@ -136,11 +142,4 @@ private void updateButtons() {
         ButterKnife.apply(buttons, VISIBILITY, View.VISIBLE);
         pickButton.setVisibility(View.GONE);
     }
-
-    static final ButterKnife.Setter<View, Integer> VISIBILITY = new ButterKnife.Setter<View, Integer>() {
-        @Override
-        public void set(final View view, final Integer visibility, int index) {
-            view.setVisibility(visibility);
-        }
-    };
```
Comment (eveliotc) [inline scissors-sample/src/main/java/com/lyft/android/scissorssample/MainActivity.java:145]: Revert all of this

File: scissors/src/main/java/com/lyft/android/scissors/CropViewExtensions.java [hunk 7]
```diff
@@ -132,43 +176,4 @@ public CropRequest quality(int quality) {
...
-        float scale;
-        if (sourceWidth * viewportHeight > viewportWidth * sourceHeight) {
-            scale = (float) viewportHeight / (float) sourceHeight;
-        } else {
-            scale = (float) viewportWidth / (float) sourceWidth;
-        }
-        final int recommendedWidth = (int) ((sourceWidth * scale) + 0.5f);
-        final int recommendedHeight = (int) ((sourceHeight * scale) + 0.5f);
-        return new Rect(0, 0, recommendedWidth, recommendedHeight);
-    }
```
Comment (eveliotc) [inline scissors/src/main/java/com/lyft/android/scissors/CropViewExtensions.java:200]: Revert this

File: scissors/src/main/java/com/lyft/android/scissors/UILBitmapLoader.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    @Override
+    public void load(@Nullable Object model, @NonNull ImageView view) {
+        final DisplayImageOptions options = new DisplayImageOptions.Builder()
+                .cacheInMemory(false)
+                .cacheOnDisk(false)
+                .displayer(bitmapDisplayer)
+                .build();
+
+        if (model instanceof String || model == null) {
+            imageLoader.displayImage((String) model, view, options);
```
Comment (eveliotc) [inline scissors/src/main/java/com/lyft/android/scissors/UILBitmapLoader.java:44]: Does Android-Universal-Image-Loader only support `String` models?

File: scissors/src/main/java/com/lyft/android/scissors/UILBitmapLoader.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.widget.ImageView;
+
+import com.nostra13.universalimageloader.core.DisplayImageOptions;
+import com.nostra13.universalimageloader.core.ImageLoader;
+import com.nostra13.universalimageloader.core.display.BitmapDisplayer;
+
+public class UILBitmapLoader implements BitmapLoader {
```
Comment (eveliotc) [inline scissors/src/main/java/com/lyft/android/scissors/UILBitmapLoader.java:17]: Document this please

File: scissors/src/main/java/com/lyft/android/scissors/UILFillViewportDisplayer.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package com.lyft.android.scissors;
+
+import android.graphics.Bitmap;
+import android.graphics.Rect;
+
+import com.nostra13.universalimageloader.core.assist.LoadedFrom;
+import com.nostra13.universalimageloader.core.display.BitmapDisplayer;
+import com.nostra13.universalimageloader.core.imageaware.ImageAware;
+
+public class UILFillViewportDisplayer implements BitmapDisplayer {
```
Comment (eveliotc) [inline scissors/src/main/java/com/lyft/android/scissors/UILFillViewportDisplayer.java]: Make this package private

File: scissors/build.gradle [hunk 11]
```diff
@@ -32,6 +32,7 @@ dependencies {
     provided 'com.squareup.picasso:picasso:[2.4.0, 2.5.2)'
     // Glide
     provided 'com.github.bumptech.glide:glide:[3.5.0, 3.6.1)'
+    provided 'com.nostra13.universalimageloader:universal-image-loader:1.9.4'
```
Comment (eveliotc) [inline scissors/build.gradle]: Please update proguard accordingly, also allow to use certain range of versions e.g  `[1.x.x, 1.9.4)`

**Comment by eveliotc** [comment]
@skyfishjy thanks for this, please use https://github.com/lyft/java-code-styles

I will add it to contribution guidelines.

File: scissors/src/main/java/com/lyft/android/scissors/UILBitmapLoader.java (hunk 8 above)
Comment (skyfishjy) [inline scissors/src/main/java/com/lyft/android/scissors/UILBitmapLoader.java:44]: Yes

**Comment by eveliotc** [comment]
@skyfishjy can you please also add proper proguard rules?

**Comment by eveliotc** [comment]
@skyfishjy cool please resolve your conflicts, also just added 

https://github.com/lyft/scissors/blob/master/CONTRIBUTING.md

Please make sure of signing our CLA

https://github.com/lyft/CLA

**Comment by skyfishjy** [comment]
Done

**Comment by eveliotc** [comment]
@skyfishjy awesome, just please squash to a single commit and we are ready to go!

**Comment by skyfishjy** [comment]
Done

**Comment by eveliotc** [comment]
@skyfishjy did you push? still seeing as 8 commits

screen shot 2015-12-02 at 7 14 05 pm

**Comment by skyfishjy** [comment]
Done

**Pull request merged**